core/unix/signal.c

```diff
@@ -2293,17 +2293,29 @@ check_signals_pending(dcontext_t *dcontext, thread_sig_info_t *info)
 /* Returns whether to execute the syscall */
 bool
 handle_sigprocmask(dcontext_t *dcontext, int how, kernel_sigset_t *app_set,
-                   kernel_sigset_t *oset, size_t sigsetsize)
+                   kernel_sigset_t *oset, size_t sigsetsize, uint *errno_val)
 {
     thread_sig_info_t *info = (thread_sig_info_t *)dcontext->signal_field;
     int i;
-    kernel_sigset_t safe_set;
+    kernel_sigset_t safe_set, safe_old_set;
+    if ((app_set != NULL && !d_r_safe_read(app_set, sizeof(safe_set), &safe_set)) ||
+        (oset != NULL && !d_r_safe_read(oset, sizeof(safe_old_set), &safe_old_set))) {
+        if (errno_val != NULL)
+            *errno_val = EFAULT;
+        return false;
+    }
+    if (app_set != NULL && !(how >= SIG_BLOCK && how <= SIG_SETMASK)) {
+        if (errno_val != NULL)
+            *errno_val = EINVAL;
+        return false;
+    }
```

Comment:
What about an invalid sigsetsize

Reply:
Will need to research a bit more about what kernel expects in the size arg. I'm leaning towards submitting this fix for a bad address first, so that the immediate issue is resolved. Added a TODO to check bad size arg too.

Reply:
I think we have to return EINVAL for anything but the size we're using now since we won't properly handle it anyway.  We can delay handling some other size if the kernel supports some other size (doubt it).

Reply:
I'm seeing something weird with MacOS and the size. sigprocmask on MacOS also uses the same path. However, sigprocmask does NOT have a 4th arg for sigsetsize. The current code was anyway extracting the 4th arg and using it as size. When I changed it to use sizeof(sigset_t) for mac, other test started failing (signal0000, etc). So that existing bug needs to be fixed.

Reply:
Discussed that Mac and others are all using kernel_sigset_t size and ignoring the passed in size. For non-Mac, made changes to check that the passed-in size is each to sizeof(kernel_sigset_t).

Reply:
The sigaction test is still failing on Mac. I tried tmate to debug it on the GA runner: it seems that the pre and post syscall handling isn't even being invoked for sigprocmask. I reverted the mac related changes in test. But keeping the changes in os.c and signal.c code. It still performs the same thing as before for mac, uses sizeof(kernel_sigset_t) for the safe read, and an additional safe-write check with the same size.

Reply:
I created i#5256 for this issue on mac.